### PR TITLE
Add waveform extractor `save()` function and Zarr backend

### LINKED ISSUE #1058: [feat-core] Add save() function to WaveformExtractor
When an existing `WaveformExtractor` is instantiated, it should be possible to `save()` it (including extensions) to a new folder.
(or to a zarr backend)

### PR DESCRIPTION
**Pull request by alejoe91**
Implements  and closes #1058

### CODE REVIEW

File: spikeinterface/core/waveform_extractor.py [hunk 1]
```diff
@@ -566,14 +765,21 @@ def get_waveforms(self, unit_id, with_index=False, cache=False, memmap=True, spa
...
+                    if memmap:
+                        wfs = np.load(str(waveform_file), mmap_mode="r")
+                    else:
+                        wfs = np.load(waveform_file)
+                elif self.format == "zarr":
+                    waveforms_group = self._waveforms_root["waveforms"]
+                    if f'waveforms_{unit_id}' not in waveforms_group.keys():
+                        raise Exception('Waveforms not extracted yet: '
+                                        'please do WaveformExtractor.run_extract_waveforms() first')
+                    wfs = waveforms_group[f'waveforms_{unit_id}'][:]
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:782]: add `memmap` option here

File: spikeinterface/core/waveform_extractor.py [hunk 2]
```diff
@@ -1031,11 +1265,23 @@ def __init__(self, waveform_extractor):
         
         if self.waveform_extractor.folder is not None:
             self.folder = self.waveform_extractor.folder
-            self.extension_folder = self.folder / self.extension_name
-
-            if not self.extension_folder.is_dir():
-                self.extension_folder.mkdir()
+            if not self.waveform_extractor.folder.suffix == ".zarr":
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:1268]: use `self.waveform_extractor.format`

File: spikeinterface/core/waveform_extractor.py [hunk 3]
```diff
@@ -1044,6 +1290,40 @@ def __init__(self, waveform_extractor):
...
+        assert 'params' in extension_group.attrs, f'No params file in extension {cls.extension_name} folder'
+        params = extension_group.attrs['params']
+
+        if waveform_extractor is None:
+            waveform_extractor = WaveformExtractor.load(folder)
+        
+        # make instance with params
+        ext = cls(waveform_extractor)
+        ext._params = params
+        ext._load_extension_data()
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:1323]: move to `load()`

File: spikeinterface/core/waveform_extractor.py [hunk 4]
```diff
@@ -1071,57 +1351,111 @@ def load_from_folder(cls, folder, waveform_extractor=None):
...
+                import numcodecs
+
+                compressor = kwargs.get("compressor", None)
+                if compressor is None:
+                    compressor = get_default_zarr_compressor()
+                for ext_data_name, ext_data in self._extension_data.items():
+                    if ext_data_name in self.extension_group.keys():
+                        del self.extension_group[ext_data_name]
+                    if isinstance(ext_data, dict):
+                        self.extension_group.create_dataset(name=ext_data_name, data=[ext_data],
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:1427]: use attribute?

File: spikeinterface/core/waveform_extractor.py [hunk 5]
```diff
@@ -1071,57 +1351,111 @@ def load_from_folder(cls, folder, waveform_extractor=None):
...
+                                                            object_codec=numcodecs.JSON())
+                        self.extension_group.attrs["is_dict"] = True
+                    elif isinstance(ext_data, np.ndarray):
+                        self.extension_group.create_dataset(name=ext_data_name, data=ext_data,
+                                                            compressor=compressor)
+                    elif isinstance(ext_data, pd.DataFrame):
+                        ext_data_array = ext_data.to_numpy().astype("float")
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:1434]: try `xarray`

File: spikeinterface/core/waveform_extractor.py [hunk 6]
```diff
@@ -1071,57 +1351,111 @@ def load_from_folder(cls, folder, waveform_extractor=None):
...
+                ext_data_ = self.extension_group[ext_data_name]
+                if "is_dict" in ext_data_.attrs:
+                    ext_data = ext_data_[0]
+                elif "index" in ext_data_.attrs:
+                    import pandas as pd
+                    ext_data = pd.DataFrame.from_records(ext_data_, columns=ext_data_.attrs["columns"])
+                    ext_data["index"] = ext_data_.attrs["index"]
+                    ext_data.set_index("index", drop=True, inplace=True)
+                    ext_data.index.rename("", inplace=True)
+                    print(ext_data)
```
Comment (alejoe91) [inline spikeinterface/core/waveform_extractor.py:1380]: remove

File: spikeinterface/core/waveform_extractor.py (hunk 1 above)
Comment (alejoe91) [reply spikeinterface/core/waveform_extractor.py:782]: or `lazy_load`?

File: spikeinterface/core/waveform_extractor.py (hunk 5 above)
Comment (alejoe91) [reply spikeinterface/core/waveform_extractor.py]: done, it works well :)

File: spikeinterface/core/waveform_extractor.py [hunk 7]
```diff
@@ -1414,20 +1421,15 @@ def _save(self, folder=None, **kwargs):
...
-                        ext_data_array = ext_data.to_numpy().astype("float")
-                        dset = self.extension_group.create_dataset(name=ext_data_name, data=ext_data_array,
-                                                                   compressor=compressor)
-                        index_dict = check_json({"index": ext_data.index.values})
-                        dset.attrs["index"] = index_dict["index"]
-                        dset.attrs["columns"] = list(ext_data.columns.values)
-                        
+                        ext_data.to_xarray().to_zarr(store=self.extension_group.store,
+                                                     group=f"{self.extension_group.name}/{ext_data_name}",
+                                                     mode="a")
```
Comment (samuelgarcia) [inline spikeinterface/core/waveform_extractor.py:1433]: cool no ?

Comment (alejoe91) [reply spikeinterface/core/waveform_extractor.py:1433]: Yep :)

File: spikeinterface/postprocessing/principal_component.py [hunk 8]
```diff
@@ -373,6 +373,7 @@ def _fit_by_channel_local(self, sparsity, n_jobs, progress_bar):
                 with open(pca_model_file, 'rb') as fid:
                     pca_models.append(pickle.load(fid))
                 pca_model_file.unlink()
+            shutil.rmtree(tmp_folder)
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/principal_component.py:377]: what is this ?

Comment (alejoe91) [reply spikeinterface/postprocessing/principal_component.py:377]: A temporary folder needed to parallelize training

**Comment by alejoe91** [comment]
@samuelgarcia ready to merge on my side

File: spikeinterface/postprocessing/tests/common_extension_tests.py [hunk 9]
```diff
@@ -94,5 +99,28 @@ def test_extension(self):
...
+                ext_data_zarr = ext_zarr._extension_data[ext_data_name]
+                if isinstance(ext_data_mem, np.ndarray):
+                    np.testing.assert_array_equal(ext_data_mem, ext_data_folder)
+                    np.testing.assert_array_equal(ext_data_mem, ext_data_zarr)
+                elif isinstance(ext_data_mem, pd.DataFrame):
+                    assert ext_data_mem.equals(ext_data_folder)
+                    assert ext_data_mem.equals(ext_data_zarr)
+                else:
+                    print(f"{ext_data_name} of type {type(ext_data_mem)} not tested.")
+
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/tests/common_extension_tests.py:126]: perfect

**Review by samuelgarcia — APPROVED** [review]

**Pull request merged**